### PR TITLE
New Addon: Fullscreen Editor

### PR DESCRIPTION
**Pull request by ErrorGamer2000**
### Changes

Added the Editor fullscreen addon, which adds a fullscreen button to the editor menu bar, and makes some slight style modifications when fullscreen is enabled, as well as optionally shrinking the stage and expanding it on fullscreen exit.

Also changed all addons that add elements to the project bar to use redux listeneres rather than click listeners.

### Tests

All changed addons hide and show as expected, and new addon does not appear to cause any conflicts.

https://user-images.githubusercontent.com/70162741/208743619-e926c3f5-4a9c-4e21-ad41-0449144d5667.mp4

### CODE REVIEW

**Comment by WorldLanguages** [comment]
Please add screenshots/video to the PR's first post :)

**Comment by ErrorGamer2000** [comment]
> Please add screenshots/video to the PR's first post :)

My apologies, updated

**Comment by WorldLanguages** [comment]
Will we need a shared space where the "full screen" icon will be inserted? Technically, "messages in editor" already uses it.

**Comment by ErrorGamer2000** [comment]
> Will we need a shared space where the "full screen" icon will be inserted? Technically, "messages in editor" already uses it.

Already added, way ahead of you there @WorldLanguages.

**Comment by fractioneater** [comment]
In the video, it doesn't look like the stage is filling all of the available space when "shrink stage" is enabled.

Also, shrinking the stage doesn't do anything when "hide stage" is enabled, so you might want to make that clear.

**Comment by ErrorGamer2000** [comment]
> In the video, it doesn't look like the stage is filling all of the available space when "shrink stage" is enabled.

That is a result of the sprite search bar, something that needs to be looked at.

> Also, shrinking the stage doesn't do anything when "hide stage" is enabled, so you might want to make that clear.

IDK

File: addon-api/content-script/Tab.js [hunk 1]
```diff
@@ -555,6 +555,11 @@ export default class Tab extends Listenable {
         },
         until: () => [],
       },
+      beforeEditorProfile: {
+        element: () => q("[class^='menu-bar_account-info-group']"),
```
Comment (Samq64) [inline addon-api/content-script/Tab.js:559]: This element only exists when signed in, so the editor fullscreen addon won't work when signed out.

File: addons/fullscreen-editor/addon.json [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  ],
+  "userstyles": [
+    {
+      "matches": ["projects"],
+      "url": "userstyle.css"
+    }
+  ],
+  "settings": [
+    {
+      "id": "shrinkStage",
```
Comment (Samq64) [inline addons/fullscreen-editor/addon.json:21]: Could this setting be hidden when the hide stage setting is enabled?

File: addons/fullscreen-editor/userscript.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  fullscreen.className = addon.tab.scratchClass("menu-bar_menu-bar-item", "menu-bar_hoverable", {
+    others: "sa-fullscreeneditor",
+  });
+
+  let isFullscreen = false;
+
+  const enterFullscreen = () => {
+    document.body.requestFullscreen();
+    if (addon.settings.get("shrinkStage")) {
+      if (!addon.settings.get("hide")) {
```
Comment (Samq64) [inline addons/fullscreen-editor/userscript.js:20]: It's possible for both settings to be enabled at the same time, even if the one to hide this stage is hidden from the settings page.

File: addons/fullscreen-editor/userstyle.css [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+.sa-fullscreeneditor::before {
+  /* Icon is in a pseudo-element so dark mode can apply a filter to it
+  without affecting the message count */
+  content: "";
+  display: block;
+  position: absolute;
+  width: 1.7rem;
+  height: 100%;
+  background: url(https://scratch.mit.edu/static/assets/422b48c2206d5240a6717c92496ba21a.svg) no-repeat center center;
```
Comment (Samq64) [inline addons/fullscreen-editor/userstyle.css:15]: Ideally both icons would be stored in the addon folder.

File: addons/mouse-pos/stage-size.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
+export const stageEventTarget = new EventTarget();
```
Comment (Samq64) [inline addons/mouse-pos/stage-size.js:1]: Is there a reason this module is in `mouse-pos`?

File: addons/fullscreen-editor/addon.json (hunk 2 above)
Comment (ErrorGamer2000) [reply addons/fullscreen-editor/addon.json:21]: Not with the way that the settings page is currently set up, I'll work on fixing that.

Comment (Samq64) [reply addons/fullscreen-editor/addon.json:21]: `"if": {"settings": {"hide": false }}`?

File: addons/mouse-pos/stage-size.js (hunk 5 above)
Comment (WorldLanguages) [reply addons/mouse-pos/stage-size.js:1]: I think it's the first addon that added its own UI to the top of the stage.

**Comment by WorldLanguages** [comment]
Let's move this to #6422

**Pull request closed without merging**